include/broker/alm/stream_transport.hh

```diff
@@ -444,211 +274,528 @@ public:
+  }
+
+  void handle_buffered_msg(caf::inbound_path*, caf::outbound_path*,
+                           caf::message& msg) {
+    // TODO: we can make this more efficient with CAF 0.18 typed views.
+    if (msg.match_elements<peer_id_list, vector_timestamp, filter_type>()) {
+      handle_filter_update(msg.get_mutable_as<peer_id_list>(0),
+                           msg.get_mutable_as<vector_timestamp>(1),
+                           msg.get_as<filter_type>(2));
+    } else if (msg.match_elements<peer_id_list, vector_timestamp, peer_id_type,
+                                  filter_type>()) {
+      handle_path_revocation(msg.get_mutable_as<peer_id_list>(0),
+                             msg.get_mutable_as<vector_timestamp>(1),
+                             msg.get_as<peer_id_type>(2),
+                             msg.get_as<filter_type>(3));
+    } else {
+      BROKER_ERROR("unexpected buffered message:" << msg);
+    }
   }
 
-  /// Pushes data to peers and stores.
-  void push(command_message msg) {
-    BROKER_TRACE(BROKER_ARG(msg));
-    remote_push(make_node_message(std::move(msg), dref().options().ttl));
-    // local_push(std::move(x), std::move(y));
+  template <class DownstreamMsg>
+  std::enable_if_t<std::is_same<typename DownstreamMsg::outer_type,
+                                caf::downstream_msg>::value>
+  handle_buffered_msg(caf::inbound_path* path, caf::outbound_path*,
+                      DownstreamMsg& msg) {
+    handle(path, msg);
   }
 
-  /// Pushes data to peers and stores.
-  void push(message_type msg) {
-    BROKER_TRACE(BROKER_ARG(msg));
-    remote_push(std::move(msg));
+  template <class UpstreamMsg>
+  std::enable_if_t<std::is_same<typename UpstreamMsg::outer_type,
+                                caf::upstream_msg>::value>
+  handle_buffered_msg(caf::inbound_path*, caf::outbound_path* path,
+                      UpstreamMsg& msg) {
+    handle(path->slots.invert(), msg);
   }
 
-  // -- communication that bypasses the streams --------------------------------
+  // -- callbacks --------------------------------------------------------------
 
-  void ship(data_message& msg, const communication_handle_type& hdl) {
-    self()->send(hdl, atom::publish_v, atom::local_v, std::move(msg));
+  /// Pushes `msg` to local workers.
+  void ship_locally(const data_message& msg) {
+    if (!worker_manager().paths().empty())
+      worker_manager().push(msg);
+    super::ship_locally(msg);
   }
 
-  template <class T>
-  void ship(T& msg) {
-    push(std::move(msg));
+  /// Pushes `msg` to local stores.
+  void ship_locally(const command_message& msg) {
+    if (!store_manager().paths().empty())
+      store_manager().push(msg);
+    super::ship_locally(msg);
   }
 
-  void ship(message_type& msg) {
-    push(std::move(msg));
+  void shutdown(shutdown_options options) {
+    BROKER_TRACE(BROKER_ARG(options));
+    if (!store_manager().paths().empty()) {
+      // TODO: honor wait-for-stores flag.
```

Comment:
I didn't see it being used in any `btest` either. I don't know whether it's important for any Zeek script/tool. I mostly left this todo since implementing graceful store shutdown (and enabling it by default) cleans up state in a cluster faster (by not having to wait on timeouts). So I don't think it's important to have right now but we should ultimately implement it as followup at some point.